subprojects/docs/src/snippets/tutorial/extraProperties/groovy/build.gradle

```diff
@@ -23,10 +23,13 @@ sourceSets {
 }
 
 tasks.register('printProperties') {
+    def springVersion = springVersion
+    def emailNotification = emailNotification
+    def productionSourceSets = sourceSets.matching { it.purpose == "production" }.collect { it.name }.toList()
```

Comment:
🔴 This changes semantics as the `sourceSets` collection will now be iterated at task realization time. Evaluation order becomes an issue as a new source-set could still be registered after the task has been realized. 

A solution would be to make the expression lazy via `providers.provider { ...}` like it was done for the `onlyIf` snippet, another would be to make `productionSourceSet` a `ListProperty<String>` and use `ListProperty#addAll`.

Reply:
This stuff is hard to get right. I will not be surprised if I let similar cases get through in my hackathon reviews.

Reply:
IIUC, the goal is to avoid evaluating at configuration action execution time, and postpone it to serialization time (still not execution), as by then all configuration actions will have been executed. 

This is what it looks like now:

```
    def productionSourceSets = provider {
        sourceSets.matching { it.purpose == "production" }.collect { it.name }
    }
    doLast {
        // ...
        productionSourceSets.get().each { println it }
    }
```